tensorflow_probability/python/experimental/stats/sample_stats.py

```diff
@@ -389,3 +393,185 @@ def finalize(self, state):
       mean: An estimate of the mean.
     """
     return state.mean
+
+
+RunningPotentialScaleReductionState = collections.namedtuple(
+    'RunningPotentialScaleReductionState', 'chain_var')
+
+
+class RunningPotentialScaleReduction(object):
+  """Holds metadata for and computes a running rhat diagnostic statistic.
+
+  `RunningPotentialScaleReduction` uses Gelman and Rubin (1992)'s potential
+  scale reduction (or rhat) for chain convergence [1]. This object also
+  supports both batching and chunking.
+
+  If multiple rhat computations are desired (batching), one should use a
+  (possibly nested) collection for initialization parameters `num_chains`
+  and `shape`. Subsequent chain states used to update the streaming rhat
+  should mimic their identical structure.
+
+  In computation, samples can be provided individually or in chunks. A
+  "chunk" of size M implies incorporating M samples into a single expectation
+  computation at once, which is more efficient than one by one. The `chunk_axis`
+  parameter in the `update` method defines chunking semantics.
+
+  `RunningPotentialScaleReduction` objects do not hold state information. That
+  information, which includes intermediate calculations, are held in a
+  `RunningPotentialScaleReductionState` as returned via `initialize` and
+  `update` method calls.
+
+  `RunningPotentialScaleReduction` is meant to serve general streaming rhat.
+  For a specialized version that fits streaming over MCMC samples, see
+  `RhatReducer` in `tfp.experimental.mcmc`.
+
+  #### References
+
+  [1]: Andrew Gelman and Donald B. Rubin. Inference from Iterative Simulation
+       Using Multiple Sequences. _Statistical Science_, 7(4):457-472, 1992.
+  """
+
+  def __init__(self, shape, num_chains, dtype=tf.float32):
+    """Instantiates this object.
+
+    Args:
+      shape: Python `Tuple` or `TensorShape` representing the shape of
+        incoming samples. Using a collection enables batching and implies
+        that future samples will mimic that exact structure.
+      num_chains: A (possibly nested) collection of integers representing
```

Comment:
??  Is this code batching over different Tensors in a list structure?  That's very unconventional in TFP; generally we batch over Tensor dimensions.

Reply:
Yeah... good point. Not entirely sure what I was thinking with this

Reply:
Ah, I remember now. My thought was that if we wanted to truly batch multiple R-hat calculations for whatever reason (as supported by `tfp.mcmc.potential_scale_reduction` -- one can imagine in a more complex way than `test_batching` in `sample_stats_test`), then you can't really pack all of that into a `Tensor` as the latents may be completely different from one another. Let me know if I started elaborating on a completely different point than the one you were trying to get at.